Hive alternative

pradovic opened the pull request:
- Hive implementation:  https://hackmd.io/xHp5c3-LQ2qIsx11wnSJQw?both.
- Hive tests
- `BroadcastPeers` instead of `BroadcastPeer` method, so we can use it to send multiple peers.
- `BroadcastPeers` implementation change: broadcast all known nodes to newly connected node and broadcast this node to all connected nodes 
- `addressbook.inmem` `Overlays` and `Multiaddresses` methods implementation, so they can be used in tests.

TODO:
- Error test cases 
- Currently, hive just adds new nodes to addressbook. We need to figure out where and who will decide and connect new nodes. Maybe we can do direct connect now, or leave this for later when Kademlia and addressbook is implemented as well.
- Addressbook implementation

acud commented on pkg/hive/hive.go:
```diff
@@ -71,80 +67,73 @@ func (s *Service) Protocol() p2p.ProtocolSpec {
-
-			if _, err := s.connecter.Connect(ctx, addr); err != nil {
-				s.logger.Infof("Connect failed for %s: %w", addr.String(), err)
+	go func() {
```
consider the following for an even further simplification (where even this `Init` would not be needed at all):
1. When topology driver broadcasts every connected peer _to_ a newly added peer
2. When topology driver broadcasts every newly connected peer _to_ every connected peer (everyone else)

acud replied:
i'd like to add something that came into mind, is that we _might_ need a message exchange to ask for peers, because if we go with an approach of just normal gossiping - we'd indeed need to keep a queue of peers we'd be able to connect to when saturation depth decreases (but this is a kademlia implementation detail. it might not infer a message exchange at all)

pradovic replied:
Makes sense, let's ee when it comes, and we can add it. There is no problem adding another stream for fetching the peers to a protocol, and it will work, along this "broadcast" style of receiving peers, no problem.

pradovic commented on pkg/addressbook/inmem/inmem.go:
```diff
@@ -45,3 +44,21 @@ func (i *inmem) Put(overlay swarm.Address, addr ma.Multiaddr) (exists bool) {
 	return e
 }
+
+func (i *inmem) Keys() []swarm.Address {
```
Sure, I generally was thinking to either export an interface and have more meaningful names, or just keep it low-level, to be used in tests. I decided for the latter, but the first approach is also fine for me, I will switch to that.

pradovic commented on pkg/discovery/mock/mock.go:
```diff
@@ -5,34 +5,36 @@
-	d.records[addressee.String()] = broadcastRecord{overlay: overlay, multiaddr: addr}
+func (d *Discovery) BroadcastPeers(addressee swarm.Address, peers ...discovery.BroadcastRecord) error {
+	for _, peer := range peers {
+		d.mtx.Lock()
```
Hmmm, it depends what we consider better? Since this is mock, I though it would be ok to get more granularity on the account of performance, since it is not really overly optimized anyway. What do you think?

janos replied:
That is fine in case of the mock.

pradovic commented on pkg/hive/hive.go:
```diff
@@ -0,0 +1,127 @@
+}
+
+func (s *Service) BroadcastPeers(addressee swarm.Address, peers ...discovery.BroadcastRecord) error {
+	ctx := context.Background()
```
I agree, did not want to change `BroadcastPeer` too much.

pradovic commented on pkg/hive/hive_test.go:
```diff
@@ -0,0 +1,207 @@
+	var records []discovery.BroadcastRecord
+	var wantMsgs []pb.Peers
+
+	for i := 0; i < 100; i++ {
```
I will use batch_size multiplier and add comment.

janos commented on pkg/hive/hive_test.go:
```diff
@@ -0,0 +1,207 @@
+	var wantMsgs []pb.Peers
+
+	for i := 0; i < 100; i++ {
+		ma, err := ma.NewMultiaddr("/ip4/127.0.0.1/udp/" + strconv.Itoa(rand.Intn(1000)+1000))
```
I am sorry, I have forgot that these ports are only formal part in tests and that no listening will be done. All fine.

pradovic commented:
@janos Thanks for the review, I think I have addressed all of the comments.

janos commented on pkg/hive/export_test.go:
```diff
@@ -0,0 +1,3 @@
+package hive
```
Add copyright header.

acud commented on pkg/hive/hive.go:
```diff
@@ -0,0 +1,126 @@
+}
+
+func (s *Service) BroadcastPeers(ctx context.Context, addressee swarm.Address, peers ...discovery.BroadcastRecord) error {
+	for len(peers) > maxBatchSize {
```
can be shortened to one for loop:
```go
        max := maxBatchSize
	for len(peers) > 0 {
                if max > len(peers) { max = len(peers) }
		if err := s.sendPeers(ctx, addressee, peers[:max]); err != nil {
			return err
		}

		peers = peers[max:]
	}
```

pradovic replied:
I agree, nice. It does have an additional if statement in the for loop and an additional variable, but it does look more compact. I like both solutions, so I can change if you like this one more?

pradovic replied:
Changed for the shorter loop.

pradovic commented on pkg/hive/hive.go:
```diff
@@ -0,0 +1,126 @@
+		}
+
+		// todo: this might be changed depending on where do we decide to connect peers
+		s.addressBook.Put(swarm.NewAddress(newPeer.Overlay), addr)
```
I think that we agreed on meetings that if we would connect each time that it would not be long-term, but maybe good only for MVP. Plus, we talked about deciding which peers  to connect to and which peers to  to recommend to other peers. I think, in order to acomplish this we would need to inform some structure/service (topology driver, or address book, or something else) that would make a decision and also keep track of connected and known peers. The first, mvp, simple implementation of this service could be to just connect to each peer, but I don't think that connecting from hive is a good idea. @janos, @acud let me know what you think.

acud replied:
@pradovic and I thought we agreed that we move towards a minimal implementation that's supposed to be deployed on a cluster. This iteration does not fulfill this expectation, and I thought we agreed that the topology driver will be able to decide which peer to connect to. It must not be through a `Connect` method (whose semantics necessarily suggests that a connection the requested peer would be made), but maybe through an `AddPeer` method that would subsequently _maybe_ connect to the peer then add the peer to the address book.

I'm not sure what do you mean by developing a "king of connection tracker" - that is the topology driver in itself, so I'm not sure about introducing more synthetic abstractions here. I would definitely not want another moving part in creating a topology (i.e. considering and testing kademlia-to-come implementation which is complex enough with another cog that is some independent component that decides whether to connect to peers according to some other intricate unbeknownst logic - there's no need for semantic differentiation between the "connection tracker" and the topology driver).

pradovic replied:
I agree it can be done in topology driver, as I wrote in the comment. Some kind of connection tracker is just a phrase here (no kings mentioned anywhere), my point was just not to use a `Connect` method directly. So, adding the `AddPeer` method to a topology driver sounds right to me.
Another point is that I thought that address book should contain addresses of known but not connected peers as well, so they can be used to try new connections when number of connections drops. That is why I thought that we should add peer to an address book in each case, although I do agree it could be done by topology driver as well.

pradovic replied:
Also current implementation of the full connectivity topology driver `AddPeer` is called from within `libp2p.Connect` method, and assumes that the address is already in the address book. It seemed to me that it's point was to broadcast the nodes, other then connect or/and add addresses to the address book. That is why I was leaning to another abstraction as I did not want to change the implementation too much.

pradovic replied:
Pls check the last commit and lmnwyt.

pradovic commented on pkg/hive/hive_test.go:
```diff
@@ -0,0 +1,217 @@
+	var records []discovery.BroadcastRecord
+	var wantMsgs []pb.Peers
+
+	// 2*maxBatchSize is ok to be used in the 2 batches corner-case test
```
I would like to add a few more test cases to these tests. Error cases + corner cases related to  batch size to test the batching logic. So, the maximum number of random resources (swarm.Addresses, multiaddrs, etc..) would be twice the size of a batch size to test this corner cases (2*maxBatchSize, 2*maxBatchSize-1). Does this make sense?

pradovic replied:
I have added these test cases and removed the comment.

pradovic commented on pkg/hive/hive_test.go:
```diff
@@ -0,0 +1,217 @@
+
+		multiaddrs = append(multiaddrs, ma)
+		addrs = append(addrs, swarm.NewAddress(createRandomBytes()))
+		if i%hive.MaxBatchSize == 0 {
```
My reasoning was that a caller of `hive.BroadcastPeers` could provide any nymber of peers for broadcast. Based on the book, and previous hive implementation I put the maximum batch size for hive to be 50. So basically, if you checkout the implementation of this function, there will be a message per batch. And single batch can contain up to 50 addresses. This is why in this place I need to create a new array (slice) of peers for each batch, so I can append elements to it, to avoid nil pointer exception at ln 60.

pradovic replied:
Done, let me know if it's better

pradovic commented:
@acud, @janos, pls take a look at the last commit. This is a try to address @acud 's comments about the connect and topology driver. I have added topology driver in hive, and removed topology driver from libp2p2, because the connection is actually being made from topology driver instead now. I have removed the topology driver from libp2p tests, and this kind of test would go to `full_tests, to test number of broadcasts. Let me know if this idea is good, and I would add tests for topology driver in the separate PR.

pradovic commented:
@acud There is a problem there then, because topology driver expect overlay address as well, even though it is connecting, which makes sense to me, because it might not be connecting stuff but save the address for later use in which case overlay address might be needed. This is why I wanted to completely separate adding known addresses and actual connecting.

pradovic commented:
@acud, I talked with @janos a little bit, and we had couple of ideas of how topology driver -> hive -> address book could be related and implemented. I think it would be best to revert my last change, and merge this PR like that (with address book put in hive), and then open a new PR working on these components and their relationship, as I think it is not that trivial. The point is that hive would probably remain 95% same no matter what we do with those components. Is this ok for as well? If it is, I will revert the last change, and we can merge PR, and then let's brainstorm a bit about how discovery, libp2p and address book should work together, to cover all use-cases nicely.
